Lombiq.ChartJs.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs

```diff
@@ -0,0 +1,147 @@
+using System;
+using System.Drawing.Imaging;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Threading.Tasks;
+// This is because both of SixLabors.ImageSharp and System.Drawing are contains Image class,
+// but we want it from SixLabors.ImageSharp
+using Bitmap = System.Drawing.Bitmap;
+
+namespace Lombiq.ChartJs.Tests.UI.Extensions;
+public static class TestCaseUITestContextExtensions
+{
+    public static async Task TestChartJsSampleBehaviorAsync(this UITestContext context)
+    {
+        await context.SignInDirectlyAsync();
+        await context.ExecuteChartJsSampleRecipeDirectlyAsync();
+        await context.TestChartJsBarChartAsync();
+        await context.TestChartJsLineChartAsync();
+    }
+
+    public static async Task TestChartJsBarChartAsync(this UITestContext context)
+    {
+        await context.GoToBalanceAsync();
+
+        context.TestChartJsChart("BarChart", "bar_chart", 5);
+    }
+
+    public static async Task TestChartJsLineChartAsync(this UITestContext context)
+    {
+        await context.GoToHistoryAsync();
+
+        context.TestChartJsChart("LineChart", "line_chart", 10);
+    }
+
+    private static void TestChartJsChart(
+        this UITestContext context,
+        string logHeader,
+        string referenceResourceName,
+        double pixelErrorThreshold)
+    {
+        // This is to avoid Chart.js animation related issues
+        var hash = context.WaitChartJsCanvasToBeReadyAndHash();
+        hash.ShouldNotBeNullOrEmpty();
+        context.Scope.AtataContext.Log.Trace($"{logHeader}: imageHash: {hash}");
+        var canvas = context.Driver.FindElementByTagName("canvas");
+        canvas.ShouldNotBeNull();
+        using var canvasImage = context.TakeScreenshotImage(canvas)
+            .ToImageSharpImage();
+        canvasImage.ShouldNotBeNull()
+            .SaveAsBmp($"Temp/{logHeader}_canvas.bmp");
+        using var referenceImage = GetResourceImageSharpImage($"Lombiq.ChartJs.Tests.UI.Assets.{referenceResourceName}.dib");
+        referenceImage.ShouldNotBeNull()
+            .SaveAsBmp($"Temp/{logHeader}_reference.bmp");
+        using var diffImage = ImageSharpCompare.CalcDiffMaskImage(
+            $"Temp/{logHeader}_canvas.bmp",
+            $"Temp/{logHeader}_reference.bmp");
+        diffImage.ShouldNotBeNull()
+            .SaveAsBmp($"Temp/{logHeader}_diff.bmp");
+        var diff = ImageSharpCompare.CalcDiff(
+            $"Temp/{logHeader}_canvas.bmp",
+            $"Temp/{logHeader}_reference.bmp");
+        context.Scope.AtataContext.Log.Trace($@"{logHeader}: diff:
+    absoluteError={diff.AbsoluteError.ToString(CultureInfo.InvariantCulture)},
+    meanError={diff.MeanError.ToString(CultureInfo.InvariantCulture)},
+    pixelErrorCount={diff.PixelErrorCount.ToString(CultureInfo.InvariantCulture)},
+    pixelErrorPercentage={diff.PixelErrorPercentage.ToString(CultureInfo.InvariantCulture)}
+");
```

Comment:
Is this needed or was it just for debugging during development? If needed, can you format it to be a bit more readable?

Also log messages should be constants. Use the additional parameters to pass in values.